typescript/packages/jumble/src/contexts/AuthenticationContext.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { createContext, useCallback, useContext, useEffect, useState } from "react";
+import { Identity, KeyStore, PassKey } from "@commontools/identity";
+
+// Location in storage of root key.
+const ROOT_KEY = "$ROOT_KEY";
+// "Name" of default persona derived from root key.
+const DEFAULT_PERSONA = "default";
+
+interface AuthenticationContextType {
+  // The authenticated user/persona.
+  user: Identity | void;
+  // Set the authenticated user/persona.
+  authenticate: () => Promise<void>;
+  // Call PassKey registration.
+  registerUser: (name: string, displayName: string) => Promise<void>;
+  // Clears authentication database.
+  clearAuthentication: () => Promise<void>;
+  // Internal: Root key.
+  root: Identity | void;
+  // Internal: Persistent storage for keys. 
+  keyStore: KeyStore | void;
+}
+
+const AuthenticationContext = createContext<AuthenticationContextType>(null!);
+
+// (async) open a key store
+// (async) check to see key exists
+export const AuthenticationProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+  const [user, setUser] = useState<Identity | void>(undefined);
+  const [keyStore, setKeyStore] = useState<KeyStore | void>(undefined);
+  const [root, setRoot] = useState<Identity | void>(undefined);
+
+  // On load, open the KeyStore and find a root key.
+  useEffect(() => {
+    let ignore = false;
+    async function getKeyStoreAndRoot() {
+      let keyStore = await KeyStore.open();
+      let root = await keyStore.get(ROOT_KEY);
+      if (!ignore) {
+        setKeyStore(keyStore);
+        if (root) {
+          setRoot(root);
+        }
+      }
+    }
+    getKeyStoreAndRoot(); 
+    return () => { ignore = true; }
+  }, []);
+
+  // When root changes, update `user` to the default persona
+  useEffect(() => {
+    let ignore = false;
+    async function setPersona() {
+      setUser(undefined);
+      if (!root) {
+        return;
+      }
+      let user = await root.derive(DEFAULT_PERSONA);
+      if (!ignore) {
+        setUser(user);
+      }
+    }
+    setPersona();
+    return () => { ignore = true; }
```

Comment:
Similar here, should we do `setUser(undefined)` on cleanup?

Reply:
SG, was viewing this more as a "cancel" than "cleanup", though I guess this would only be executed if the AuthenticationContext was removed from the DOM (IIUC)